### ISSUE TITLE
Array operators do not allow user code to define reflected operators.

### ISSUE THREAD

**Issue by junov** [google/jax#2001]
Use case:

A framework built on top of JAX defines an array-like class with binary operators that allow it to be api-compatible with jax's DeviceArray

```
class NewThingy():

  def __add__(self, device_array):
   ...
  def __radd__(self, device_array):
   ...
```

FYI: `__radd__` is the reflected version of `__add__`, which is meant to be called when the second operand of the "+" operator is a NewThingy.

Currently this does not work because JAX does not respect the standard way of implementing type handling in binary operators.  In the case where the left operand is a DeviceArray and the right operand is a NewThingy, the python interpreter first tries to call `DeviceArray.__add__`, and if the method cannot handle the other operand (because it is of an incompatible type), then the interpreter will try the reflected operator provided by `NewThingy.__radd__`.  For this to work, `DeviceArray.__add__` needs to return NotImplemented when the second operand is incompatible.  Instead, the current implementation raises  a TypeError which prevents Python's operator reflection mechanism from kicking in.

To fix this, we would need the functions created by _one_to_one_binop and _maybe_bool_binop (in numpy/lax_numpy.py) to return NotImplemented whenever _check_arraylike detects an incompatible argument.

**Comment by shoyer** [comment]
This should be fixed by https://github.com/google/jax/pull/1942

**Comment by mattjj** [comment]
Now I understand #1942 better! Sounds like it's worth merging (as if @shoyer wasn't convincing enough on his own).

**Comment by mattjj** [comment]
Since we merged #1942, I think we can close this now. Please reopen if I'm mistaken!
